### ISSUE TITLE
Ignore vs raise exception?

### ISSUE THREAD

**Issue by prlaba** [scottctr/NStateManager#17]
I'm just getting familiar with NStateManager.  Very nice code and features.  Well done!

This package appears similar to (derived from?) the Stateless package.

One thing I noticed in your package is that there is no easy way to specify which triggers (a.k.a. events) are allowed or not allowed for a given state in the configuration.  If a trigger is fired while in a given state, and that trigger is not included in any .AddTransition or .AddTriggerAction in the state's configuration, no action is taken and the trigger is simply ignored.  That may be the desired behavior in some state machines, but not in the one I'm designing. 

I discovered this playing around with the Sale Console example, by firing a ChangeGiven trigger while in the Open state (that trigger should only be fired while in the ChangeDue state).  The program'a output was identical with or without my extra firing of the ChangeGiven trigger, with no indication that an attempt had been made to fire that trigger.  No harm was done, since the trigger was ignored, but the fact that I was able to fire a trigger while in a 'wrong' state is my concern.

The Stateless package takes a different approach, where you must include a Permit (or PermitIf) method for each allowed trigger within a given state's configuration.  Without a Permit method for the ChangeGiven trigger in the Open state's configuration, Stateless would have thrown an exception in the above situation.

Is there another way within the state manager's Configure method, to indicate which triggers are allowed or not allowed for a given state?

**Comment by scottctr** [comment]
Hey @prlaba.

You've found one of the gaps I need to fill. It's on the todo list (#5), but just haven't addressed it yet. Now that you've brought it up, I'll knock it out this weekend. I'll probably just expose an Action (OnNoTransitionFound?) you can assign to do whatever you need.

In the mean time, there's an easy workaround. FireTrigger returns a [result](https://github.com/scottctr/NStateManager/blob/master/Source/NStateManager/StateTransitionResult.cs) that tells you a few things about what occurred while processing the trigger. That result can be used to detect when a transition wasn't defined for the current/state:
```C#
var result = stateMachine.FireTrigger(....);
if (result.StartingState.CompareTo(result.PreviousState) == 0
    && result.PreviousState.CompareTo(result.CurrentState) == 0)
{
    // log it, throw exception, etc..
}
```
That check may be a bit of overkill, but it's the safest way until I complete some testing. I'll give you an update once I get this in place.

Keep the questions and suggestions coming!

**Comment by prlaba** [comment]
Thanks Scott.  Yes, I was aware that the FireTrigger methods returned a StateTransitionResult object, but it never occurred to me to use that as a way to detect a trigger that didn't result in a state transition.  I'm still learning.

I like your suggestion of using a configuration method to define a user-supplied method to be called if a trigger is fired for a given state that is not 'included' in that state's configuration.

But just to be clear, this method should *not* be called simply because a trigger doesn't cause a state transition (even a reentry into the same state) -- there's several valid ways that can happen, such as an AddDynamicTransition where the transition's condition is false.  The user-supplied method should only be called if a trigger is fired while in a given state and that state's configuration doesn't include that trigger in any of its AddTriggerAction, AddTransition, or AddDynamicTransition configuration methods.  Basically, if a trigger isn't 'mentioned' in a state's configuration, the user-supplied method should be called whenever the trigger is fired while in that state.

Re the name of the new method, how about OnOtherTriggers?  ('OtherTriggers' referes to those triggers not included in a state's configuration.)  That seems to nicely convey its purpose and usage.

Thanks again,

Paul

**Comment by prlaba** [comment]
Perfect, having both methods covers my needs as well as an easy way to handle non-transitioning triggers.

And I like your method names too, well describes their meaning.

I’m looking forward to checking them out!

Thanks again,

Paul

**Comment by prlaba** [comment]
Scott, I wanted to offer a slightly different approach to resolving this issue, one that would not require any new syntax.  This is the approach taken by the Appcelerate.StateMachine package.

If a Trigger is fired while in a state that has no transition defined for that trigger, raise an UnhandledTrigger event (C# event, not state machine event), passing with it the details (currrnt state & trigger that was fired).  I can then subscribe to that event if I need to handle such cases.

Appcelerate.StateMachine calls this event TransitionDeclined. You could use this instead, although it’s meaning would extend to any situation where a trigger was fired but no transition occurred.

This has the advantage of preserving the current StateMachine behavior (ignoring fired triggers that have no defined transitions in the current state), while still allowing a client to know about such situations.  And it doesn’t require any new syntax.

Paul

**Comment by scottctr** [comment]
I'm open to different approaches, but I don't think I'm following you on this one yet. UnhandledTrigger sounds like the OnTriggerNotConfigured I mention above since it will include a StateTransitionResult with the details you mention. Whatever the name, this functionality should not affect the state machine's flow **unless** the handler raises an exception.

Please let me know what I'm missing. Seeing your thoughts in code might help to clarify so feel free to fork the code or just post some sample code here.

**Comment by scottctr** [comment]
The confusion is on me, Paul. I should have provided a few more details of what I was thinking. The good news is that that we're on the same page. 

I finished up the automatic transitions today so I should be able to put the new events in place and get a new build out tomorrow.

**Comment by scottctr** [comment]
OK, I've uploaded the latest with these new events. It's also published on NuGet as v1.2.1.
